File: 백준/Silver/17124. 두 개의 배열/두 개의 배열.py
import sys
input = sys.stdin.readline

T = int(input())

for t in range(T):
    n, m = map(int, input().split())
    A = list(map(int, input().split()))
    B = list(map(int, input().split()))
    C = []
    B.sort()
    sum = 0

    for num in A:
        closeNum = B[0]
        minDiff = abs(num - closeNum)
        lt = 0
        rt = m - 1

        while lt <= rt:
            mid = (lt + rt) // 2

            if B[mid] == num:
                closeNum = num
                break
            elif B[mid] > num:
                rt = mid - 1
            else:
                lt = mid + 1

            cal = abs(num - B[mid])

            if minDiff > cal or (minDiff == cal and closeNum > B[mid]):
                minDiff = cal
                closeNum = B[mid]

        sum += closeNum

    print(sum)

File: 백준/Silver/17124. 두 개의 배열/두 개의 배열.java
import java.util.*;
import java.io.*;

class Main {
    public static void main(String[] args) throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        StringBuilder sb = new StringBuilder();

        int T = Integer.parseInt(br.readLine());
        
        for (int tc = 0; tc < T; tc++) {
            st = new StringTokenizer(br.readLine());
            int n = Integer.parseInt(st.nextToken());
            int m = Integer.parseInt(st.nextToken());

            int[] A = new int[n];
            int[] B = new int[m];
            long answer = 0;
            st = new StringTokenizer(br.readLine());
            for (int i = 0; i < n; i++) {
                A[i] = Integer.parseInt(st.nextToken());
            }
            st = new StringTokenizer(br.readLine());
            for (int i = 0; i < m; i++) {
                B[i] = Integer.parseInt(st.nextToken());
            }

            Arrays.sort(B);
            
            for (int i = 0; i < n; i++) {
                int closeNum = B[0];
                int minDiff = Math.abs(A[i] - closeNum);
                int lt = 0;
                int rt = m-1;
                while(lt <= rt) {
                    int mid = (lt + rt) / 2;
    
                    if (A[i] == B[mid]) {
                        closeNum = B[mid];
                        break;
                    } else if (A[i] < B[mid]) {
                        rt = mid - 1;
                    } else {
                        lt = mid + 1;
                    }

                    int cal = Math.abs(A[i]-B[mid]);
                    
                    if (cal < minDiff || (cal == minDiff && B[mid] < closeNum)) {
                        minDiff = cal;
                        closeNum = B[mid];
                    }

                    
                }
                answer += closeNum;
            }
            sb.append(answer).append("\n");
        }
        System.out.println(sb);
        
    }
}